[TCFMM-3] Add specification

dcastro opened the pull request:
[//]: # (This is a template of a pull request template.)
[//]: # (You should modify it considering specifics of a particular repository and put it there.)
[//]: # (Comments like this are meta-comments, they shouldn't be present in the final template.)
[//]: # (Comments starting with '<!---' are intended to stay in the final template.)

[//]: # (Keep in mind that it's only a template which contains items relevant to almost all conceivable repository.)
[//]: # (There can be other important items relevant to your repository that you can add here.)

## Description

<!--
Describes the nature of your changes. If they are substantial, you should
further subdivide this into a section describing the problem you are solving and
another describing your solution.
-->

[//]: # (Here you can add a link to the corresponding issue tracker, e. g. https://issues.serokell.io/issue/AD-)
[//]: # (For GitHub/GitLab issues it is better to use just hash symbol or exclamation mark as it is more resistant to repo movements)
[//]: # (In this case please also prefix the link with "Resolves" keyword)
[//]: # (See https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
[//]: # (See https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords)
## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

Resolves TCFMM-3

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

[//]: # (Update link to style guide if necesary or remove if it's not present)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

Martoon-00 commented on docs/specification.md:
```diff
@@ -0,0 +1,596 @@
+Conceptually, at every block level, the contract will calculate a cumulative sum
+of the _current tick index_ `ic` since the contract's inception.
+
+At any point in time `t`, the acumulator `tick_cumulative(t)` is equal to:
```
Since I'm reading the text, just some edits along the way

```suggestion
At any point in time `t`, the accumulator `tick_cumulative(t)` is equal to:
```

Martoon-00 commented on docs/specification.md:
```diff
@@ -0,0 +1,596 @@
+- If the value of `max_amount_x` and/or `max_amount_y` is `0` no `transfer` will
+  be made for the respective token.
+- If the accrued fees are more than the `max_amount_x` and/or `max_amount_y`
+  the amount `transfer`red will be capped for the respective token.
```
Would be nice to have a motivation why it should work exactly this way, if it's known.

dcastro replied:
In our original design, the user would specify exactly how many tokens he wanted to withdraw. If they tried to withdraw more tokens than are available, the operation would fail.

However, because it's hard to know precisely how many tokens can be withdrawn at a given point in time, it would be easier to let the user specify a maximum ceiling of tokens to withdraw. If there are fewer tokens available, then they're all withdrawn.

This means the user can simply call `transfer_protocol_fees` with `max_amount_x=999999999...` to withdrawn all available tokens.

Martoon-00 replied:
Yes, for this particular case it is definitely convenient. 

But maybe the user will want to have a way to submit a precise number of tokens (supposedly much fewer than the currently collected amount of fees is). One use case - when the recipient of the portion of the fee is some other person's address, which maybe be useful to shortcut the money transfer and avoid paying extra Tezos fees.

dcastro replied:
It appears this entrypoint won't be needed after all :sweat_smile:

Martoon-00 replied:
:arrow_right: :arrow_right: :arrow_right: :wastebasket:

Martoon-00 reviewed (commented):
Cool, thanks! This document provides a lot of missing context to me

dcastro commented:
@Martoon-00 I've updated the spec with the notes from last meeting, please review the last 4 fixups.

Summary of the fixups:
1. explained how the procotol fees work and under which conditions are they applied
2. deleted `maximum_liquidity_present` field, removed admin entrypoints, updated entrypoints' docs regarding protocol fees.
3. added docs for the `increase_observation_count` entrypoint
4. fixed some links (maybe we could extend xrefcheck's [copy-paste protection checks](https://github.com/serokell/xrefcheck/issues/64) to scan "reference-style" links as well :smile: )

Martoon-00 commented:
> to scan "reference-style" links as well 

Oh that sounds interesting, I'll mention this use case.

Martoon-00 commented on docs/specification.md:
```diff
@@ -0,0 +1,555 @@
+  * `seconds_per_liquidity_cumulative` is the cumulative seconds per
+    liquidity-in-range at that `timestamp`.
+- If any of the given timestamps is too far back in the past,
+  fails with `INVALID_TIMESTAMP`.
```
I think it's worth mentioning the motivation for storing a finite number of observations right here. That is a reasonable question that will probably arise in the mind of anyone who's reading this point.

AFAIU we were asked to avoid `big_map`s because they are costly, and generally adding some info to storage after each update in the contract is expensive for Tezos nodes?

dcastro replied:
Done

Martoon-00 reviewed (commented):
Leaving some more comments, but overall looks very good :+1:

Martoon-00 commented on docs/specification.md:
```diff
@@ -0,0 +1,555 @@
+### **Set_position**
+
+Updates or creates a new [position](#positions) in the given range.
+
```
It may be worth mentioning, that changing position by 0 is a valid way to obtain the current fees from it.

dcastro replied:
Done

dcastro commented on docs/specification.md:
```diff
@@ -0,0 +1,555 @@
+The greater the number, the further back in time users will be able to
+retrieve data using [`observe`][#observe].
+
+The caller of this entrypoint will pay for the additional storage costs.
```
>  I suppose we won't just initialize `new observation_count - old observation_count`` slots with dummy values, or we will?

Haven't thought too much about it yet, but yes, that sounds like the most probable scenario.

Circular buffers are usually implemented by having a fixed-size array, and two pointers to where the actual contents begin and end (`lo` and hi`), and making sure `lo` and `hi` never cross each other. It doesn't matter what's stored in the slots above `hi` or below `lo`, because reading those values would be considered an "out of bounds access".

In our case, the fixed size array will be a big_map emulating a linked list. But, apart from that, I think the usual strategy should work the same.

Martoon-00 replied:
Thanks, your explanation sounds reasonable.

I guess later, if the first version of the implementation suddenly turns out to diverge from the spec here, let's consider this issue there.

dcastro commented on docs/specification.md:
```diff
@@ -0,0 +1,555 @@
+- If the swap is no longer acceptable because the `deadline` was not met, fails
+  with `TIMED_OUT`.
+- If less than `min_dy` amount of token `Y` would be obtained from the swap, fails
+  with `UNDER_MIN`.
```
> This sounds like we fail with a string, but in fact, we are going to fail with nat, right? Or do you want these sentences to be updated in #8?

Correct! Depending on which PR is merged first (#8 or #4), either me or @rinn7e will have to update these references (and update the `Errors` section at the bottom with links to @rinn7e 's files).

> Also, I think TIMED_OUT error should eventually include the deadline and the actual execution time to make the failure reason clearer for the users. Simiarly for UNDER_MIN. What do you think?

Agreed! In general, I believe the errors should contain more contextual info.

This is a bit annoying to do in ligo though, due to the [limitations of `failwith`](https://gitlab.com/ligolang/ligo/-/issues/193), and so we need to use the "embedded michelson" trick :/ But still, yes, we should do that.

Martoon-00 replied:
Nice!

> This is a bit annoying to do in ligo though, due to the limitations of failwith, and so we need to use the "embedded michelson" trick :/ But still, yes, we should do that.

Yeah, fully agree. Hope they finally decide to implement this soon :roll_eyes:

dcastro commented on docs/specification.md:
````diff
@@ -0,0 +1,555 @@
+    min_dx : nat ;
+    to_dx : address ;
+}
+```
````
Done

dcastro commented:
@Martoon-00 I believe I've addressed your last batch of comments, please have another look :)

Martoon-00 commented:
Thanks for addressing those points, looks good!
